rfcs/33-apps_own_scheduling.md

```diff
@@ -0,0 +1,239 @@
+
+### Code organization
+
+As plugins no longer depend on `App` information at all, they should be moved into `bevy_ecs` directly.
+
+### System sets
+
+The changes proposed are simple ergonomic renaming to reflect the increased prominence of system sets.
+`App::add_system_set` should be changed to `App::add_systems`, and `SystemSet::with_system` should be shortened to `SystemSet::with`.
+
+### Resources
+
+All resources are added to the app before the schedule begins. The app should panic if duplicate resources are detected at this stage; app users can manually remove them from one or more plugins in a transparent fashion to remove the conflict if needed.
+
+### Schedule overhaul prerequisites
+
+1. Stage boundaries must be weakened. This API assumes that e.g. system sets can span hard sync points to allow them to be inserted into a given state and that ordering dependencies operate across hard sync points.
+2. All system scheduling information must be stored in the system descriptor (which are then stored within system sets) in order to enable a unified approach to configuration. See [Bevy #2736](https://github.com/bevyengine/bevy/pull/2736) for a nearly complete PR on this point.
+3. Users must be able to configure systems by reference to a particular label. This proposal has previously been called **label properties**.
+
+### Default stages and runner
+
+If plugins can no longer configure the `App` in arbitrary ways, we need a new, ergonomic way to set up the default schedule and runner for standard Bevy games.
+
+The best way to do this is to create two new builder methods on `App`: `App::minimal` and `App::default`, which sets up the default stages, sets the runner and adds the required plugins.
+
+### Plugin groups
+
+Plugin groups now require the simple `IntoPluginGroup` trait with a `into` method that returns a `Vec<Plugin>`.
+
+Alternatively, these could just be completely deprecated due to the move towards `App::default`, which was overwhelmingly their most common use.
+
+## Drawbacks
+
+1. Plugins are no longer quite as powerful as they were. This is both good and bad: it reduces their expressivity, but reduces the risk that they mysteriously break your app.
+2. Apps can directly manipulate the system ordering of their plugins. This is essential for ensuring that interoperability and configurability issues are resolved, but risks breaking internal guarantees.
+3. Users can accidentally break the requirement for a hard sync point to pass between systems within a plugin as encouraged by the plugin configuration by moving system sets into the same stage. Hard ordering dependencies briefly discussed in the *Future Work* section of this RFC could ensure that this is impossible.
+4. Replacing existing resource values is more verbose, requiring either a startup system or explicitly removing the resource from the pre-existing plugin. This is much more explicit, but could slow down some common use cases such as setting window size.
```

Comment:
Could we have some kind of plugin resource (name bikeshedding necessary) concept. Plugins would then specify some data and the plugin resource will be called when plugins are added and will handle this data appropriately? For example in case of the TypeRegistry plugin resource a plugin can provide type registrations, while for the AssetServer plugin resource a plugin can provide asset loaders. This would possibly also allow handling removal of plugins using an extra callback in the future.

Reply:
I don't fully follow this. Could you walk through the "order of operations" of how the AssetPlugin registers the AssetServer and some CustomPlugin registers a custom AssetLoader? How would the AssetLoader be represented in the CustomPlugin? When / how would it be registered with the AssetServer resource?

Reply:
AssetPlugin would register AssetServer as "plugin resource" using something like `self.add_plugin_resource(AssetServer::default())`. CustomPlugin schedules an add event for CustomAssetLoader for the AssetServer "plugin resource" using something like `self.plugin_resource_add_part::<AssetServer>(CustomAssetLoader::default())`. The order between the these two registrations doesn't matter. The `AssetServer` could then either be notified about the add of `CustomAssetLoader` after all plugins are registered or as soon as both `AssetServer` and `CustomAssetLoader` are registered. It doesn't really matter which one is chosen. Notably in either case if the `AssetServer` is not yet registered by the time `CustomPlugin` tries to register `CustomAssetLoader`, the actual registration is delayed until after `AssetServer` is registered. `AssetServer` has to implement a trait with a method that is called every time something is registered with it.

Reply:
Yup that seems like it would work. This seems similar to the "Plugins don't have dependencies" solution (due to the need to ensure that AssetServer is registered prior to applying "resource parts"), but with more complicated userspace (due to the addition of "resource parts" as a concept). It means that in addition to the "normal" apis exposed on a resource, developers also need to think about the "resource plugin api" and expose the relevant bits. Feels like a bit too much complexity. If we can detect the `resource_part -> resource` dependency, then we can just as easily provide the custom asset loader registration with direct (but deferred) access to that resource without the need for a new public api layer.

Reply:
I guess so. It doesn't help with removing plugins again though for the editor or runtime loadable mods.